core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -117,11 +120,10 @@ public ThisT set(String property, String value) {
     return self();
   }
 
-  protected PartitionSpec writeSpec() {
-    Preconditions.checkState(spec != null,
-        "Cannot determine partition spec: no data or delete files have been added");
+  protected PartitionSpec dataSpec() {
```

Comment:
Renaming this does require touching more places but I think keeping it `writeSpec` will be confusing.